### ISSUE TITLE
PHP_ROUND_* rounding mode consts not recognised

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#4058]
Consts presents since 5.3 not recognised by psalm, instead interpreted as class references.

* PHP_ROUND_HALF_UP
* PHP_ROUND_HALF_DOWN
* PHP_ROUND_HALF_EVEN
* PHP_ROUND_HALF_ODD

https://psalm.dev/r/ba85d72e07

**Comment by SignpostMarv** [comment]
have tried defining them in stub files, but doesn't want to seem to work?

**Comment by SignpostMarv** [comment]
popping them into ConstFetchAnalyzer also does not seem to work :s

**Comment by weirdan** [comment]
Ah, you can't use standalone constants as types (as opposed to class constants). Psalm cannot distinguish them syntactically, as PHP allows classes (and functions) with the same name: https://3v4l.org/AJgv9

**Comment by SignpostMarv** [comment]
> Psalm cannot distinguish them syntactically, as PHP allows classes (and functions) with the same name

I refer you to this variant, where php-parser *should* pick up the const import statements: https://psalm.dev/r/1b7b200d4d which would be able to indicate to psalm that something is a const.

**Comment by SignpostMarv** [comment]
> Ah, you can't use standalone constants as types (as opposed to class constants).

I also refer you to this variant, where no error occurs when it should: https://psalm.dev/r/9fadf4b1b8

**Comment by SignpostMarv** [comment]
poking around things further, it looks like `TypeExpander::expandUnion` switches the param type from TInt to TMixed ?

**Comment by SignpostMarv** [comment]
re: this example:
```php
<?php

class Foo {
	const ROUND_HALF_UP = \PHP_ROUND_HALF_UP;
	const ROUND_HALF_DOWN = \PHP_ROUND_HALF_DOWN;
	const ROUND_HALF_EVEN = \PHP_ROUND_HALF_EVEN;
	const ROUND_HALF_ODD = \PHP_ROUND_HALF_ODD;
}

/**
* @param Foo::ROUND_HALF_UP $mode
*/
function foo(int $mode) : void {
    echo (string) $mode;
}

bar(5);

```

this is what I've got working so far- it stuffs the resolved type in, but FunctionLikeAnalyzer doesn't pick up on it.
```diff
diff --git a/src/Psalm/Internal/Analyzer/ClassAnalyzer.php b/src/Psalm/Internal/Analyzer/ClassAnalyzer.php
index 4ba15cb97..a3e71dbff 100644
--- a/src/Psalm/Internal/Analyzer/ClassAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/ClassAnalyzer.php
@@ -913,6 +913,29 @@ class ClassAnalyzer extends ClassLikeAnalyzer
 
                     $trait_file_analyzer->clearSourceBeforeDestruction();
                 }
+            } elseif ($stmt instanceof PhpParser\Node\Stmt\ClassConst) {
+                if (
+                    $const->value instanceof PhpParser\Node\Expr\ConstFetch
+                ) {
+                    $const_name = implode('\\', $const->value->name->parts);
+
+                    $const_type = Statements\Expression\Fetch\ConstFetchAnalyzer::getConstType(
+                        $statements_analyzer,
+                        implode('\\', $const->value->name->parts),
+                        $const->value->name instanceof PhpParser\Node\Name\FullyQualified,
+                        $class_context
+                    );
+
+                    if ($const_type) {
+                        if ($stmt->isProtected()) {
+                            $storage->protected_class_constants[$const_name] = $const_type;
+                        } elseif ($stmt->isPrivate()) {
+                            $storage->private_class_constants[$const_name] = $const_type;
+                        } else {
+                            $storage->public_class_constants[$const_name] = $const_type;
+                        }
+                    }
+                }
             }
         }
```

**Comment by SignpostMarv** [comment]
@weirdan it's also worth noting that this example behaves as expected, the only difference being that it's not using the consts (for the reasons you mentioned) https://psalm.dev/r/b45369de02

**Comment by weirdan** [comment]
> this is what I've got working so far- it stuffs the resolved type in, but FunctionLikeAnalyzer doesn't pick up on it.

I thought I fixed this issue in #4065 but apparently the fix was plain wrong, as your example still fails to fail properly in that branch.

> I refer you to this variant, where php-parser should pick up the const import statements: [snip] which would be able to indicate to psalm that something is a const.

I'm not sure if you're figured this out already, but php-parser can't help here because it's totally valid to have a constant **and** a class (and a function to boot) with the same name in PHP, in the same namespace. PHP itself disambiguates symbols by how and where they are used, and stores them in separate symbol tables. This is why you have to refer to class names as `SomeClass::class` and not just `SomeClass`, also this is why you have separate `use`, `use const` and `use function`. These are all syntactically different specifically to allow disambiguation. Thus, `/** @param ABC $p */` (if ABC here is allowed to refer to non-class constants as well) is ambiguous - you might be referring to a `const ABC` or `class ABC` here, and there's no way to tell them apart.

> I also refer you to this variant, where no error occurs when it should

Yep, I intend to fix that in #4065

**Comment by SignpostMarv** [comment]
> I'm not sure if you're figured this out already, but php-parser can't help here because it's totally valid to have a constant **and** a class (and a function to boot) with the same name in PHP, in the same namespace.

I see this frequently as i'm working on legacy projects, but as of php 5.6 we can `use const Foo;` which should signal developer intent that `Foo` should be referred to as a const- [this technique](https://github.com/vimeo/psalm/issues/4058#issuecomment-680793188) serves as a workaround for the disambiguation in the meantime.

> also this is why you have separate `use`, `use const` and `use function`
> Thus, `/** @param ABC $p */` (if ABC here is allowed to refer to non-class constants as well) is ambiguous - you might be referring to a `const ABC` or `class ABC` here, and there's no way to tell them apart.

see also developer intent of `use const PHP_ROUND_HALF_UP;` should resolve to the const, but `use PHP_ROUND_HALF_UP;` should not.


> Yep, I intend to fix that in #4065

👍
